File: Android/app/src/main/java/ru/inwords/inwords/data/sync/TranslationSyncController.java
package ru.inwords.inwords.data.sync;

import android.util.Log;
import io.reactivex.Completable;
import io.reactivex.Observable;
import io.reactivex.Single;
import io.reactivex.observables.GroupedObservable;
import io.reactivex.schedulers.Schedulers;
import io.reactivex.subjects.PublishSubject;
import ru.inwords.inwords.core.util.SchedulersFacade;
import ru.inwords.inwords.dagger.annotations.CacheRepository;
import ru.inwords.inwords.dagger.annotations.LocalRepository;
import ru.inwords.inwords.data.dto.EntityIdentificator;
import ru.inwords.inwords.data.dto.WordTranslation;
import ru.inwords.inwords.data.repository.translation.TranslationWordsLocalRepository;
import ru.inwords.inwords.data.repository.translation.TranslationWordsRemoteRepository;
import ru.inwords.inwords.domain.util.WordsUtilKt;

import javax.inject.Inject;
import java.util.Arrays;
import java.util.Collection;
import java.util.List;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;

public class TranslationSyncController {
    private final TranslationWordsLocalRepository inMemoryRepository;
    private final TranslationWordsLocalRepository localRepository;
    private final TranslationWordsRemoteRepository remoteRepository;

    private final AtomicInteger dataChangesCounter;
    private final PublishSubject<Integer> dataChangedNotifier;

    @Inject
    TranslationSyncController(
            @CacheRepository TranslationWordsLocalRepository inMemoryRepository,
            @LocalRepository TranslationWordsLocalRepository localRepository,
            TranslationWordsRemoteRepository remoteRepository) {
        this.inMemoryRepository = inMemoryRepository;
        this.localRepository = localRepository;
        this.remoteRepository = remoteRepository;

        this.dataChangedNotifier = PublishSubject.create();
        this.dataChangesCounter = new AtomicInteger();

        establishSyncAllReposWithCacheWatcher();
    }

    enum Groups {
        REMOVE_REMOTE, REMOVE_LOCAL, ADD, NORMAL
    }

    @SuppressWarnings("ResultOfMethodCallIgnored")
    private void establishSyncAllReposWithCacheWatcher() {
        dataChangedNotifier
                .debounce(2, TimeUnit.SECONDS)
                .doOnNext(o -> trySyncAllReposWithCache()
                        .subscribeOn(Schedulers.io())
                        .subscribe(() -> {
                        }, t -> Log.e("TranslationSync", "" + t.getMessage())))
                .subscribe();
    }

    public void notifyDataChanged() {
        dataChangedNotifier.onNext(dataChangesCounter.incrementAndGet());
    }

    public Single<PullWordsAnswer> presyncOnStart() {
        return localRepository.getList()
                .flatMapSingle(inMemoryRepository::addReplaceAll)
                .flatMap(Observable::fromIterable)
                .map(EntityIdentificator::getServerId)
                .filter(serverId -> serverId != 0)
                .toList()
                .doOnError(t -> Log.e(this.getClass().getSimpleName(), "" + t.getMessage()))
                .flatMap(remoteRepository::pullWords)
                .doOnError(t -> Log.e(this.getClass().getSimpleName(), "" + t.getMessage()))
                .doOnSuccess(pullWordsAnswer -> {
                    List<Integer> removedServerIds = pullWordsAnswer.getRemovedServerIds();
                    List<WordTranslation> addedWords = pullWordsAnswer.getAddedWords();

                    if (!removedServerIds.isEmpty()) {  //Its IMPORTANT to remove before addReplace because
                        //its important
                        Throwable throwable = Completable.mergeDelayError(
                                Arrays.asList(
                                        localRepository.removeAllServerIds(removedServerIds),
                                        inMemoryRepository.removeAllServerIds(removedServerIds)
                                ))
                                .blockingGet();

                        if (throwable != null) {
                            Log.e(this.getClass().getSimpleName(), "" + throwable.getMessage());
                        }
                    }

                    if (!addedWords.isEmpty()) {
                        Single.mergeDelayError(
                                localRepository.addReplaceAll(addedWords),
                                inMemoryRepository.addReplaceAll(addedWords))
                                .blockingSubscribe(wordTranslations -> {
                                }, t -> Log.e(this.getClass().getSimpleName(), "" + t.getMessage()));
                    }
                })
                .subscribeOn(SchedulersFacade.INSTANCE.io());
    }

    public Completable trySyncAllReposWithCache() {
        return inMemoryRepository.getList()
                .observeOn(SchedulersFacade.INSTANCE.computation())
                .firstElement() //Берём все элементы только 1 раз
                .flatMapObservable(Observable::fromIterable) //Выдаём их по одному
                .groupBy(this::group) //Группируем
                .flatMapSingle(GroupedObservable::toList) //Каждую группу пихаем в List
                .filter(wordTranslations -> !wordTranslations.isEmpty()) //Смотрим, чтобы он был не пустой
                .observeOn(SchedulersFacade.INSTANCE.io())
                .flatMapCompletable(this::groupedListHandler);
    }

    private Groups group(WordTranslation wordTranslation) {
        int serverId = wordTranslation.getWordIdentificator().getServerId();

        if (serverId == 0) {
            return Groups.ADD;
        } else if (wordTranslation.isLocallyDeleted()) {
            return Groups.REMOVE_LOCAL;
        } else if (wordTranslation.isRemoteDeleted()) {
            return Groups.REMOVE_REMOTE;
        }

        return Groups.NORMAL;
    }

    private Completable groupedListHandler(List<WordTranslation> list) {
        if (list.isEmpty()) {
            return Completable.complete();
        }

        switch (group(list.get(0))) { //Узнаём какой группе принадлежит лист
            case ADD: {
                return localRepository
                        .addReplaceAll(list)
                        .flatMapCompletable(wordTranslations -> remoteRepository
                                .addAll(wordTranslations)
                                .flatMapCompletable(wordIdentificatorsRemote -> {
                                    mergeIds(wordTranslations, wordIdentificatorsRemote);

                                    return Single.mergeDelayError(
                                            localRepository.addReplaceAll(wordTranslations),
                                            inMemoryRepository.addReplaceAll(wordTranslations)).ignoreElements();

                                })
                                .onErrorResumeNext(__ -> inMemoryRepository.addReplaceAll(wordTranslations).ignoreElement())
                        );
            }

            case REMOVE_REMOTE: {
                List<Integer> serverIds = WordsUtilKt.serverIdsFromWordTranslations(list);
                return remoteRepository.removeAllServerIds(serverIds)
                        .andThen(Completable.mergeDelayError(Arrays.asList(
                                inMemoryRepository.removeAllServerIds(serverIds),
                                localRepository.removeAllServerIds(serverIds))));
            }

            case REMOVE_LOCAL: {
                return Completable.mergeDelayError(Arrays.asList(
                        inMemoryRepository.removeAll(list),
                        localRepository.removeAll(list)));
            }

            case NORMAL:
            default:
                return Completable.complete();
        }
    }

    private void mergeIds(Collection<? extends WordTranslation> list, Collection<? extends EntityIdentificator> listIds) {
        if (list.isEmpty() || listIds.isEmpty()) {
            return;
        }

        for (WordTranslation wordTranslation : list) {
            for (EntityIdentificator entityIdentificator : listIds) {
                if (entityIdentificator.getId() == wordTranslation.getId()) {
                    wordTranslation.setServerId((entityIdentificator.getServerId()));
                }
            }
        }
    }

}

File: Android/app/src/main/java/ru/inwords/inwords/data/sync/TranslationSyncController.kt
package ru.inwords.inwords.data.sync

import android.util.Log
import io.reactivex.Completable
import io.reactivex.Observable
import io.reactivex.Single
import io.reactivex.schedulers.Schedulers
import io.reactivex.subjects.PublishSubject
import ru.inwords.inwords.core.util.SchedulersFacade
import ru.inwords.inwords.dagger.annotations.LocalRepository
import ru.inwords.inwords.data.dto.*
import ru.inwords.inwords.data.repository.translation.TranslationWordsLocalRepository
import ru.inwords.inwords.data.repository.translation.TranslationWordsRemoteRepository
import ru.inwords.inwords.domain.util.serverIdsFromWordTranslations
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicInteger
import javax.inject.Inject

class TranslationSyncController @Inject
internal constructor(
        @param:LocalRepository private val localRepository: TranslationWordsLocalRepository,
        private val remoteRepository: TranslationWordsRemoteRepository) {

    private val dataChangesCounter: AtomicInteger = AtomicInteger()
    private val dataChangedNotifier: PublishSubject<Int> = PublishSubject.create()

    init {
        establishSyncAllReposWithCacheWatcher()
    }

    internal enum class Groups {
        REMOVE_REMOTE, REMOVE_LOCAL, ADD, NORMAL
    }

    private fun establishSyncAllReposWithCacheWatcher() {
        dataChangedNotifier
                .debounce(2, TimeUnit.SECONDS)
                .doOnNext {
                    trySyncRemoteReposWithLocal()
                            .subscribeOn(Schedulers.io())
                            .subscribe({ }, { t -> Log.e("TranslationSync", "" + t.message.orEmpty()) })
                }
                .subscribe()
    }

    fun notifyDataChanged() {
        dataChangedNotifier.onNext(dataChangesCounter.incrementAndGet())
    }

    fun presyncOnStart(): Single<PullWordsAnswer> {
        return localRepository.getList().firstOrError() //TODO
                .map { wordTranslations ->
                    wordTranslations.filter { it.serverId != 0 }.map { it.serverId }
                }
                .doOnError { t -> Log.e(this.javaClass.simpleName, t.message.orEmpty()) }
                .flatMap { remoteRepository.pullWords(it) }
                .doOnError { t -> Log.e(this.javaClass.simpleName, t.message.orEmpty()) }
                .doOnSuccess { pullWordsAnswer ->
                    val removedServerIds = pullWordsAnswer.removedServerIds
                    val addedWords = pullWordsAnswer.addedWords

                    if (removedServerIds.isNotEmpty()) {
                        //Its IMPORTANT to remove before addReplace because its important
                        localRepository.removeAllServerIds(removedServerIds).blockingGet()
                                ?.let {
                                    Log.e(this.javaClass.simpleName, it.message.orEmpty())
                                }
                    }

                    if (addedWords.isNotEmpty()) {
                        localRepository.addReplaceAll(addedWords)
                                .doOnError { Log.e(this.javaClass.simpleName, it.message.orEmpty()) }
                                .onErrorReturnItem(emptyList())
                                .blockingGet()
                    }
                }
                .subscribeOn(SchedulersFacade.io())
    }

    fun trySyncRemoteReposWithLocal(): Completable {
        return localRepository.getList()
                .observeOn(SchedulersFacade.computation())
                .firstElement() //Берём все элементы только 1 раз
                .flatMapObservable { Observable.fromIterable(it) } //Выдаём их по одному
                .groupBy { group(it) } //Группируем
                .flatMapSingle { it.toList() } //Каждую группу пихаем в List
                .filter { wordTranslations -> wordTranslations.isNotEmpty() } //Смотрим, чтобы он был не пустой
                .observeOn(SchedulersFacade.io())
                .flatMapCompletable { groupedListHandler(it) }
    }

    private fun group(wordTranslation: WordTranslation): Groups {
        val serverId = wordTranslation.serverId

        return when {
            serverId == 0 -> Groups.ADD
            wordTranslation.isLocallyDeleted -> Groups.REMOVE_LOCAL
            wordTranslation.isRemoteDeleted -> Groups.REMOVE_REMOTE
            else -> Groups.NORMAL
        }
    }

    private fun groupedListHandler(list: List<WordTranslation>): Completable {
        if (list.isEmpty()) {
            return Completable.complete()
        }

        return when (group(list[0])) {
            //Узнаём какой группе принадлежит лист
            Groups.ADD -> {
                remoteRepository
                        .addAll(list)
                        .flatMapCompletable { wordIdentificatorsRemote ->
                            localRepository.addReplaceAll(mergeIds(list, wordIdentificatorsRemote)).ignoreElement()
                        }
                        .doOnError { t -> Log.e(this.javaClass.simpleName, t.message.orEmpty()) }
                        .onErrorComplete() //TODO

            }

            Groups.REMOVE_REMOTE -> {
                val serverIds = serverIdsFromWordTranslations(list)
                remoteRepository.removeAllServerIds(serverIds)
                        .andThen(localRepository.removeAllServerIds(serverIds))
            }

            Groups.REMOVE_LOCAL -> {
                localRepository.removeAll(list)
            }

            Groups.NORMAL -> {
                Completable.complete()
            }
        }
    }

    private fun mergeIds(list: List<WordTranslation>, listIds: List<EntityIdentificator>): List<WordTranslation> {
        if (list.isEmpty()) {
            return emptyList()
        }

        if (listIds.isEmpty()) {
            return list
        }

        val newList = mutableListOf<WordTranslation>()

        list.forEach { wordTranslation ->
            listIds.find { it.id == wordTranslation.id }?.let { newList.add(wordTranslation.withServerId(it.serverId)) }
        }

        return newList
    }

}
